File: src/main/java/datastructures/queue/java/CustomQueueCircularArray.java
package datastructures.queue.java;

import java.util.Arrays;

public class CustomQueueCircularArray<T> {

    private Object [] dataArray;
    private int size;
    private int count;
    private int head;
    private int tail;

    public CustomQueueCircularArray(int size) {
        this.dataArray = new Object[size];
        this.size = size;
    }

    public void enqueue(T element){
        if(isFull()) throw new IllegalStateException();
        dataArray[tail] = element;
        tail = (tail + 1) % size;
        count++;
    }

    public void dequeue(){
        if(count == 0) throw new IllegalStateException();
        dataArray[head] = null;
        head = (head + 1) % size;
        count--;
    }

    public Object peek(){
        return dataArray[head];
    }

    public boolean isFull() { return count == size;}

    public boolean isEmpty() { return count == 0;}

    public void display(){
        System.out.println(Arrays.toString(dataArray));
    }
}

File: src/main/java/datastructures/queue/kotlin/CustomQueueCircularArray.kt
package datastructures.queue.kotlin

class CustomQueueCircularArray<T>(val size: Int) {
    var dataArray: Array<Any?> = arrayOfNulls<Any?>(size)
    var count: Int = 0;
    var head: Int = 0;
    var tail: Int = 0;

    fun enqueue(element: T) {
        if (isFull()) throw IllegalStateException()
        dataArray[tail] = element
        tail = (tail + 1) % size
        count++
    }

    fun dequeue() {
        if (isEmpty()) throw IllegalStateException()
        dataArray[head] = null
        head = (head + 1) % size
        count--
    }

    fun peek(): Any? {
        if (isEmpty()) throw IllegalStateException()
        return dataArray[head]
    }

    fun isEmpty(): Boolean {
        return count == 0
    }

    fun isFull(): Boolean {
        return count == size
    }
}